### ISSUE TITLE
Numerical orbitals with pyscf

### ISSUE THREAD

**Issue by pulkin** [sunqm/pyscf#59]
I am having difficulties with using numerical orbitals with pyscf. I guess the actual problem is with having a node of the radial function at zero. Following is the basis set I tried:
`{'B': [[0, [1.2920777646002564, 0.4007600499432814], [0.21013549673963738, -0.8278733419943876]], [1, [10.0, 0.7138459438580154], [0.22753255448886953, -0.7606134086048335]]], 'N': [[0, [3.3158593887033017, 0.7322562299222847], [0.45416089000619037, -1.4522661009821747]], [1, [10.0, 1.7699270395553452], [0.8089497974448853, -1.881765082412774]]]}`
The band structures I obtained with this basis set using [the test file](https://github.com/sunqm/pyscf/blob/master/pbc/dft/test/test_ks_noimport.py) are almost non-dispersive. It seems like I am missing something very important here. Any suggestions?

**Comment by pulkin** [comment]
It seems like I was missing `r^{n-1}` factor.

**Comment by pulkin** [comment]
The numerical orbitals `psi(r)` work together with numerical pseudopotentials `V_eff(r)`, both defined on a grid of the radial coordinate r. It is relatively easy to expand `psi(r)` into gaussians and use these functions in pyscf, however, it is more complicated with pseudopotentials. Is it enough to implement something like `pyscf.pbc.gto.pseudo.parse_cp2k._parse`? I think that some documentation would be really helpful there.

**Comment by sunqm** [comment]
Do you need to access the pseudo potential in pyscf? I'm not clear what is your need. Would you please describe more specifically the problem you got?

**Comment by gkc1000** [comment]
Hi Artem,

PySCF has two kinds of pseudopotentials -- the HGH pseudopotential, which
is the one used in CP2K, which is of separable (Kleinman-Bylander) form,
but is expressed in terms of Gaussians; and the general ECP/BFD
pseudopotential, which is non-separable. In principle it should be possible
to fit your numerical pseudopotential (which is probably separable) to the
expansion form of the HGH pseudopotential, and then to read it in using
parse_cp2k.

Garnet

**Comment by sunqm** [comment]
Maybe you want to feed a customized PP to the code?  If so, I think a quick way is to format the PP in the internal format.  You can find the internal format here for ECP/BFD https://github.com/sunqm/pyscf/blob/master/gto/mole.py#L303  and for HGH https://github.com/sunqm/pyscf/blob/master/pbc/gto/cell.py#L46 .  ECP/BFD integrals are computed in pyscf/gto/ecp.py  , HGH PP are handled in pyscf/pbc/pseudo/*.py .

If none of them matches the format of your PP, we can create a new format.  Implementing a parser is not hard, I think we can do it later when we decide how to organize the new PP format.

**Comment by pulkin** [comment]
I think that is fine: the pseudopotentials are semi-local as far as I understand (i.e. there is simply a table with local and l-(j-)dependent components, see [link](http://www.jaist.ac.jp/~t-ozaki/vps_pao2013/H/H_CA13.vps)). I am currently building a proof-of-concept to make it all work. However, my concern is actually the following:

1. The pseudoatomic orbitals have nodes. I tried brute-force approach (least squares) to expand them into a sum of gaussians but this requires a particularly large number (tens) of gaussians, compared to nodeless lowest energy orbitals. Is it ok to have that large expansions or it will take too much time to compute Hamiltonian matrix elements? In OpenMX they do it on a real-space grid (did not check that yet).

2. Same is valid for pseudopotentials. I did not plot the components yet but if I have 10 gaussians in my orbital and 10 gaussians in my pseudopotential than I will require roughly 100x time to compute the Hamiltonian. Is it a bottleneck? If so, maybe it is better to proceed directly with OpenMX implementation from scratch. I expected to find some interfaces declaring a list of routines to implement but I did not find any. Maybe it is a good idea to have some abstract pseudopotential class and several realizations of it.
